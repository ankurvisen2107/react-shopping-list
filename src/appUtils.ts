import {Category, Item} from "./appModels";
import {ALL_CATEGORIES} from "./appConstants";

export function filterItems(items : Item[], query: string, categoryId: number) {
    const filterItems = items.filter(item =>
        item.name.toLowerCase().startsWith(query.toLowerCase())
    );
    if (categoryId>0){
        return filterItems.filter(filterItems => filterItems.category_id == categoryId);
    }
    return filterItems;
}

export function getAllCategories(categories: Category[]): Category[] {
        const allCategories = [...categories];
        allCategories.unshift(ALL_CATEGORIES);
        return allCategories;
}
