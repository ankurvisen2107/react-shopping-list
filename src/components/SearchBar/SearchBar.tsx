import React, {ChangeEventHandler, FC} from 'react';
import {FaBars, FaCheck, FaRegBell} from "react-icons/fa";
import IconPanel from "../Generic/IconPanel";
import {FaRegShareFromSquare} from "react-icons/fa6";

interface Props {
    query: string,
    onChange: ChangeEventHandler
}

const SearchBar: FC<Props> = ({query, onChange}) =>
    (
        <div className='item-row top-padding center-align'>
            <div className='item-column' style={{flexGrow: 1}}>
                <IconPanel
                    icon={<FaBars size={'1.5em'}/>}
                />
            </div>
            <div className='flex-half item-column' style={{flexGrow: 6}}>
                <input
                    type="text"
                    placeholder="Search..."
                    value={query}
                    onChange={onChange}
                />
            </div>
            <div className='item-column margin-left' style={{flexGrow: 1}}>
                <IconPanel
                    icon={<FaRegBell size={'1.5em'}/>}
                />
            </div>
            <div className='item-column' style={{flexGrow: 1}}>
                <IconPanel
                    icon={<FaCheck size={'1.5em'}/>}
                />
            </div>
            <div className='item-column' style={{flexGrow: 1}}>
                <IconPanel
                    icon={<FaRegShareFromSquare size={'1.5em'}/>}
                />
            </div>
        </div>
    );

export default React.memo(SearchBar);
