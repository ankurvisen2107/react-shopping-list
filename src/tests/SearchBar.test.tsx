import {render} from "@testing-library/react";
import {BrowserRouter} from "react-router-dom";
import {filterItems} from "../appUtils";
import {mockItemList, mockSearchItem} from "./testConstants";
import SearchBar from "../components/SearchBar/SearchBar";

test('Renders the landing page', () => {
    const func = jest.fn();
    render(<BrowserRouter><SearchBar query={mockSearchItem.name} onChange={func}/></BrowserRouter>);
});

describe('SearchBar component', () => {

    it('should filter items correctly on basis of serach query', function () {
        const mockFilteredValues = filterItems(mockItemList, mockSearchItem.name, mockSearchItem.category_id);
        expect(mockFilteredValues).toMatchObject([mockSearchItem]);
    });

    it('matches snapshot', () => {
        const func = jest.fn();
        const { container } =  render(<BrowserRouter><SearchBar query={mockSearchItem.name} onChange={func}/></BrowserRouter>);
        expect(container).toMatchSnapshot();
    });
});
