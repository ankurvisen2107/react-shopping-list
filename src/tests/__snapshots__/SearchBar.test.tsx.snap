// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar component matches snapshot 1`] = `
<div>
  <div
    class="item-row top-padding center-align"
  >
    <div
      class="item-column"
      style="flex-grow: 1;"
    >
      <div
        class="icon-panel"
      >
        <svg
          fill="currentColor"
          height="1.5em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1.5em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
          />
        </svg>
      </div>
    </div>
    <div
      class="item-column"
      style="flex-grow: 6;"
    >
      <input
        placeholder="Search..."
        type="text"
        value="testName2"
      />
    </div>
    <div
      class="item-column margin-left"
      style="flex-grow: 1;"
    >
      <div
        class="icon-panel"
      >
        <svg
          fill="currentColor"
          height="1.5em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1.5em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M439.39 362.29c-19.32-20.76-55.47-51.99-55.47-154.29 0-77.7-54.48-139.9-127.94-155.16V32c0-17.67-14.32-32-31.98-32s-31.98 14.33-31.98 32v20.84C118.56 68.1 64.08 130.3 64.08 208c0 102.3-36.15 133.53-55.47 154.29-6 6.45-8.66 14.16-8.61 21.71.11 16.4 12.98 32 32.1 32h383.8c19.12 0 32-15.6 32.1-32 .05-7.55-2.61-15.27-8.61-21.71zM67.53 368c21.22-27.97 44.42-74.33 44.53-159.42 0-.2-.06-.38-.06-.58 0-61.86 50.14-112 112-112s112 50.14 112 112c0 .2-.06.38-.06.58.11 85.1 23.31 131.46 44.53 159.42H67.53zM224 512c35.32 0 63.97-28.65 63.97-64H160.03c0 35.35 28.65 64 63.97 64z"
          />
        </svg>
      </div>
    </div>
    <div
      class="item-column"
      style="flex-grow: 1;"
    >
      <div
        class="icon-panel"
      >
        <svg
          fill="currentColor"
          height="1.5em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1.5em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          />
        </svg>
      </div>
    </div>
    <div
      class="item-column"
      style="flex-grow: 1;"
    >
      <div
        class="icon-panel"
      >
        <svg
          fill="currentColor"
          height="1.5em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 576 512"
          width="1.5em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M400 255.4V240 208c0-8.8-7.2-16-16-16H352 336 289.5c-50.9 0-93.9 33.5-108.3 79.6c-3.3-9.4-5.2-19.8-5.2-31.6c0-61.9 50.1-112 112-112h48 16 32c8.8 0 16-7.2 16-16V80 64.6L506 160 400 255.4zM336 240h16v48c0 17.7 14.3 32 32 32h3.7c7.9 0 15.5-2.9 21.4-8.2l139-125.1c7.6-6.8 11.9-16.5 11.9-26.7s-4.3-19.9-11.9-26.7L409.9 8.9C403.5 3.2 395.3 0 386.7 0C367.5 0 352 15.5 352 34.7V80H336 304 288c-88.4 0-160 71.6-160 160c0 60.4 34.6 99.1 63.9 120.9c5.9 4.4 11.5 8.1 16.7 11.2c4.4 2.7 8.5 4.9 11.9 6.6c3.4 1.7 6.2 3 8.2 3.9c2.2 1 4.6 1.4 7.1 1.4h2.5c9.8 0 17.8-8 17.8-17.8c0-7.8-5.3-14.7-11.6-19.5l0 0c-.4-.3-.7-.5-1.1-.8c-1.7-1.1-3.4-2.5-5-4.1c-.8-.8-1.7-1.6-2.5-2.6s-1.6-1.9-2.4-2.9c-1.8-2.5-3.5-5.3-5-8.5c-2.6-6-4.3-13.3-4.3-22.4c0-36.1 29.3-65.5 65.5-65.5H304h32zM72 32C32.2 32 0 64.2 0 104V440c0 39.8 32.2 72 72 72H408c39.8 0 72-32.2 72-72V376c0-13.3-10.7-24-24-24s-24 10.7-24 24v64c0 13.3-10.7 24-24 24H72c-13.3 0-24-10.7-24-24V104c0-13.3 10.7-24 24-24h64c13.3 0 24-10.7 24-24s-10.7-24-24-24H72z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
